// @virtuoso.dev/message-list/dist/index.js
import { Cell as c, Signal as E, map as g, filter as v, withLatestFrom as k, scan as ie, debounceTime as xe, mapTo as P, throttleTime as nn, delayWithMicrotask as Ze, onNext as xn, Action as on, useCellValue as M, useRealm as we, Realm as Lo, RealmContext as Ro, useCellValues as Co, usePublisher as Mo } from "@virtuoso.dev/gurx";
import h, { useEffect as Tn, useLayoutEffect as Vo } from "react";
import { jsx as y, jsxs as En } from "react/jsx-runtime";
function Le(e, t) {
  const n = c(e, (o) => {
    o.link(t(o), n);
  });
  return n;
}
const Be = { lvl: 0 };
function yn(e, t, n, o = Be, i = Be) {
  return { k: e, l: o, lvl: n, r: i, v: t };
}
function T(e) {
  return e === Be;
}
function Re() {
  return Be;
}
function zt(e, t) {
  if (T(e)) return Be;
  const { k: n, l: o, r: i } = e;
  if (t === n) {
    if (T(o))
      return i;
    if (T(i))
      return o;
    const [s, l] = $n(o);
    return mt(V(e, { k: s, l: wn(o), v: l }));
  }
  return t < n ? mt(V(e, { l: zt(o, t) })) : mt(V(e, { r: zt(i, t) }));
}
function he(e, t, n = "k") {
  if (T(e))
    return [Number.NEGATIVE_INFINITY, void 0];
  if (e[n] === t)
    return [e.k, e.v];
  if (e[n] < t) {
    const o = he(e.r, t, n);
    return o[0] === Number.NEGATIVE_INFINITY ? [e.k, e.v] : o;
  }
  return he(e.l, t, n);
}
function H(e, t, n) {
  return T(e) ? yn(t, n, 1) : t === e.k ? V(e, { k: t, v: n }) : t < e.k ? bn(V(e, { l: H(e.l, t, n) })) : bn(V(e, { r: H(e.r, t, n) }));
}
function Yt(e, t, n) {
  if (T(e))
    return [];
  const { k: o, v: i, l: s, r: l } = e;
  let r = [];
  return o > t && (r = r.concat(Yt(s, t, n))), o >= t && o <= n && r.push({ k: o, v: i }), o <= n && (r = r.concat(Yt(l, t, n))), r;
}
function Ao(e, t, n, o) {
  if (T(e))
    return Be;
  let i = Re();
  for (const { k: s, v: l } of de(e))
    s > t && s <= n ? i = H(i, ...o(s, l)) : i = H(i, s, l);
  return i;
}
function Oo(e, t, n) {
  let o = Re(), i = -1;
  for (const { start: s, end: l, value: r } of Do(e))
    s < t ? (o = H(o, s, r), i = r) : s > t + n ? o = H(o, s - n, r) : l >= t + n && i !== r && (o = H(o, t, r));
  return o;
}
function de(e) {
  return T(e) ? [] : [...de(e.l), { k: e.k, v: e.v }, ...de(e.r)];
}
function $n(e) {
  return T(e.r) ? [e.k, e.v] : $n(e.r);
}
function wn(e) {
  return T(e.r) ? e.l : mt(V(e, { r: wn(e.r) }));
}
function V(e, t) {
  return yn(t.k ?? e.k, t.v ?? e.v, t.lvl ?? e.lvl, t.l ?? e.l, t.r ?? e.r);
}
function Ht(e) {
  return T(e) || e.lvl > e.r.lvl;
}
function bn(e) {
  return jt(Mn(e));
}
function mt(e) {
  const { l: t, r: n, lvl: o } = e;
  if (n.lvl >= o - 1 && t.lvl >= o - 1)
    return e;
  if (o > n.lvl + 1) {
    if (Ht(t))
      return Mn(V(e, { lvl: o - 1 }));
    if (!T(t) && !T(t.r))
      return V(t.r, {
        l: V(t, { r: t.r.l }),
        lvl: o,
        r: V(e, {
          l: t.r.r,
          lvl: o - 1
        })
      });
    throw new Error("Unexpected empty nodes");
  }
  if (Ht(e))
    return jt(V(e, { lvl: o - 1 }));
  if (!T(n) && !T(n.l)) {
    const i = n.l, s = Ht(i) ? n.lvl - 1 : n.lvl;
    return V(i, {
      l: V(e, {
        lvl: o - 1,
        r: i.l
      }),
      lvl: i.lvl + 1,
      r: jt(V(n, { l: i.r, lvl: s }))
    });
  }
  throw new Error("Unexpected empty nodes");
}
function Do(e) {
  return Cn(de(e));
}
function Ln(e, t, n) {
  if (T(e))
    return [];
  const o = he(e, t)[0];
  return Cn(Yt(e, o, n));
}
function Rn(e, t) {
  const n = e.length;
  if (n === 0)
    return [];
  let { index: o, value: i } = t(e[0]);
  const s = [];
  for (let l = 1; l < n; l++) {
    const { index: r, value: a } = t(e[l]);
    s.push({ end: r - 1, start: o, value: i }), o = r, i = a;
  }
  return s.push({ end: Number.POSITIVE_INFINITY, start: o, value: i }), s;
}
function Cn(e) {
  return Rn(e, ({ k: t, v: n }) => ({ index: t, value: n }));
}
function jt(e) {
  const { r: t, lvl: n } = e;
  return !T(t) && !T(t.r) && t.lvl === n && t.r.lvl === n ? V(t, { l: V(e, { r: t.l }), lvl: n + 1 }) : e;
}
function Mn(e) {
  const { l: t } = e;
  return !T(t) && t.lvl === e.lvl ? V(t, { r: V(e, { l: t.r }) }) : e;
}
function yt(e, t, n, o = 0) {
  let i = e.length - 1;
  for (; o <= i; ) {
    const s = Math.floor((o + i) / 2), l = e[s], r = n(l, t);
    if (r === 0)
      return s;
    if (r === -1) {
      if (i - o < 2)
        return s - 1;
      i = s - 1;
    } else {
      if (i === o)
        return s;
      o = s + 1;
    }
  }
  throw new Error(`Failed binary finding record in array - ${e.join(",")}, searched for ${t}`);
}
function Vn(e, t, n) {
  return e[yt(e, t, n)];
}
function No(e, t, n, o) {
  const i = yt(e, t, o), s = yt(e, n, o, i);
  return e.slice(i, s + 1);
}
function sn({ index: e }, t) {
  return t === e ? 0 : t < e ? -1 : 1;
}
function Bo({ offset: e }, t) {
  return t === e ? 0 : t < e ? -1 : 1;
}
function _o(e) {
  return { index: e.index, value: e };
}
function Fo(e, t, n, o = 0) {
  return o > 0 && (t = Math.max(t, Vn(e, o, sn).offset)), t = Math.max(0, t), Rn(No(e, t, n, Bo), _o);
}
const _e = [[], 0, 0, 0];
function Ho(e, [t, n]) {
  let o = 0, i = 0, s = 0, l = 0;
  if (n !== 0) {
    l = yt(e, n - 1, sn), s = e[l].offset;
    const a = he(t, n - 1);
    o = a[0], i = a[1], e.length && e[l].height === he(t, n)[1] && (l -= 1), e = e.slice(0, l + 1);
  } else
    e = [];
  for (const { start: r, value: a } of Ln(t, n, Number.POSITIVE_INFINITY)) {
    const u = (r - o) * i + s;
    e.push({ height: a, index: r, offset: u }), o = r, s = u, i = a;
  }
  return [e, i, s, o];
}
function Po(e) {
  const { size: t, startIndex: n, endIndex: o } = e;
  return (i) => i.start === n && (i.end === o || i.end === Number.POSITIVE_INFINITY) && i.value === t;
}
function Wo(e, t) {
  let n = T(e) ? 0 : Number.POSITIVE_INFINITY;
  for (const o of t) {
    const { size: i, startIndex: s, endIndex: l } = o;
    if (n = Math.min(n, s), T(e)) {
      e = H(e, 0, i);
      continue;
    }
    const r = Ln(e, s - 1, l + 1);
    if (r.some(Po(o)))
      continue;
    let a = !1, u = !1;
    for (const { start: p, end: d, value: I } of r)
      a ? (l >= p || i === I) && (e = zt(e, p)) : (u = I !== i, a = !0), d > l && l >= p && I !== i && (e = H(e, l + 1, I));
    u && (e = H(e, s, i));
  }
  return [e, n];
}
const tt = [Re(), 0];
function zo(e, [t, n]) {
  if (n.length > 0 && T(e) && t.length === 2) {
    const o = t[0].size, i = t[1].size;
    return [
      n.reduce((s, l) => H(H(s, l, o), l + 1, i), Re()),
      0
    ];
  }
  return Wo(e, t);
}
const fe = E();
c([]);
c([]);
c(0);
c(null);
c(Number.NaN);
const Te = c(!1), J = c(tt, (e) => {
  e.link(
    e.pipe(
      fe,
      v((t) => t.length > 0),
      k(Q),
      g(([t, n]) => zo(n, [t, []]))
    ),
    J
  );
}), Q = c(tt[0], (e) => {
  e.link(
    e.pipe(
      J,
      g(([t]) => t)
    ),
    Q
  );
}), An = c(tt[1], (e) => {
  e.link(
    e.pipe(
      J,
      g(([, t]) => t)
    ),
    An
  );
}), Ee = c(_e[1]), ge = c(_e[0]), qe = c(_e, (e) => {
  e.link(
    e.pipe(
      Q,
      k(An),
      ie(([t], [n, o]) => Ho(t, [n, o]), _e)
    ),
    qe
  ), e.link(
    e.pipe(
      qe,
      g(([, t]) => t)
    ),
    Ee
  ), e.link(
    e.pipe(
      qe,
      g(([t]) => t)
    ),
    ge
  );
}), On = c(_e[2], (e) => {
  e.link(
    e.pipe(
      qe,
      g(([, , t]) => t)
    ),
    On
  );
}), Dn = c(_e[3], (e) => {
  e.link(
    e.pipe(
      qe,
      g(([, , , t]) => t)
    ),
    Dn
  );
}), We = c(0, (e) => {
  e.link(
    e.pipe(
      e.combine(Ce, Dn, On, Ee),
      g(([t, n, o, i]) => o + (t - n) * i)
    ),
    We
  );
});
function Nn(e, t) {
  if (t.length === 0)
    return [0, 0];
  const { offset: n, index: o, height: i } = Vn(t, e, sn);
  return [i * (e - o) + n, i];
}
function Je(e, t) {
  return Nn(e, t)[0];
}
function Bn(e, t) {
  return Math.abs(e - t) < 1.01;
}
function _n() {
  return typeof navigator > "u" ? !1 : /Macintosh/i.test(navigator.userAgent) && navigator.maxTouchPoints && navigator.maxTouchPoints > 1 || /iP(ad|od|hone)/i.test(navigator.userAgent) && /WebKit/i.test(navigator.userAgent);
}
function Yo(e) {
  return !e;
}
function jo(e) {
  return e === 1 ? 1 : 1 - 2 ** (-10 * e);
}
function Ct(e = 1) {
  return (t, n) => {
    const o = n.signalInstance();
    return n.sub(t, (i) => {
      let s = e;
      function l() {
        s > 0 ? (s--, requestAnimationFrame(l)) : n.pub(o, i);
      }
      l();
    }), o;
  };
}
const Kt = "up", Pt = "down", Ko = "none", qo = {
  atBottom: !1,
  notAtBottomBecause: "NOT_SHOWING_LAST_ITEM",
  state: {
    offsetBottom: 0,
    scrollTop: 0,
    viewportHeight: 0,
    viewportWidth: 0,
    scrollHeight: 0
  }
}, Uo = 0, Go = 4;
function In(e) {
  return (t, n) => {
    const o = n.signalInstance();
    return n.sub(t, (i) => {
      e > 0 ? e-- : n.pub(o, i);
    }), o;
  };
}
c(!1);
const Fn = c(!0);
E();
const Se = c(!1), Zo = E((e) => {
  e.link(e.pipe(Fn, nn(50)), Zo);
}), Hn = c(Go), Jo = c(Uo, (e) => {
  e.link(
    e.pipe(
      e.combine(w, Jo),
      g(([t, n]) => t <= n)
    ),
    Fn
  );
}), Xe = c(!1, (e) => {
  e.link(e.pipe(w, In(1), P(!0)), Xe), e.link(e.pipe(w, In(1), P(!1), xe(100)), Xe);
}), qt = c(!1, (e) => {
  e.link(e.pipe(X, P(!0)), qt), e.link(e.pipe(X, P(!1), xe(200)), qt);
}), ln = c(!1), bt = c(
  null,
  (e) => {
    e.link(
      e.pipe(
        e.combine(
          D,
          w,
          K,
          xt,
          Hn,
          Rt,
          Vt,
          Q
        ),
        v(([, , , , , , , t]) => !T(t)),
        ie((t, [n, o, i, s, l, r]) => {
          const u = o + i - n + r > -l, p = {
            viewportWidth: s,
            viewportHeight: i,
            scrollTop: o,
            scrollHeight: n,
            listMarginTop: r
          };
          if (u) {
            let I, b;
            return o > t.state.scrollTop ? (I = "SCROLLED_DOWN", b = t.state.scrollTop - o) : (I = n === i ? "LIST_TOO_SHORT" : "SIZE_DECREASED", b = t.state.scrollTop - o || t.scrollTopDelta), {
              atBottom: !0,
              state: p,
              atBottomBecause: I,
              scrollTopDelta: b
            };
          }
          let d;
          return i < t.state.viewportHeight ? d = "VIEWPORT_HEIGHT_DECREASING" : s < t.state.viewportWidth ? d = "VIEWPORT_WIDTH_DECREASING" : o < t.state.scrollTop ? d = "SCROLLING_UPWARDS" : p.scrollHeight > t.state.scrollHeight || p.listMarginTop < t.state.listMarginTop ? t.atBottom ? d = "SIZE_INCREASED" : d = t.notAtBottomBecause : t.atBottom ? d = "NOT_FULLY_SCROLLED_TO_LAST_ITEM_BOTTOM" : d = t.notAtBottomBecause, {
            atBottom: !1,
            notAtBottomBecause: d,
            state: p
          };
        }, qo)
      ),
      bt
    ), e.link(
      e.pipe(
        bt,
        ie(
          ({ prev: t }, n) => {
            const o = !!(t && n && t.atBottom && !n.atBottom && n.notAtBottomBecause === "SIZE_INCREASED");
            return {
              prev: n,
              shouldScroll: o
            };
          },
          { prev: null, shouldScroll: !1 }
        ),
        g(({ shouldScroll: t }) => t)
      ),
      ln
    ), e.sub(
      e.pipe(
        K,
        k(bt, fn, Qe),
        v(([, , t, n]) => !t && !n),
        ie(
          (t, [n, o]) => {
            let i = 0;
            return t.viewportHeight > n && o && !o.atBottom && o.notAtBottomBecause === "VIEWPORT_HEIGHT_DECREASING" && (i = t.viewportHeight - n), { viewportHeight: n, delta: i };
          },
          { viewportHeight: 0, delta: 0 }
        )
      ),
      (t) => {
        t.delta && e.pub(X, t.delta);
      }
    );
  },
  (e, t) => !e || e.atBottom !== (t == null ? void 0 : t.atBottom) ? !1 : !e.atBottom && !t.atBottom ? e.notAtBottomBecause === t.notAtBottomBecause : !0
), Pn = c(0, (e) => {
  e.link(
    e.pipe(
      e.combine(w, D, K),
      ie(
        (t, [n, o, i]) => {
          if (!Bn(t.scrollHeight, o)) {
            const s = o - (n + i) < 1;
            return t.scrollTop !== n && s ? {
              scrollHeight: o,
              scrollTop: n,
              jump: t.scrollTop - n,
              changed: !0
            } : {
              scrollHeight: o,
              scrollTop: n,
              jump: 0,
              changed: !0
            };
          }
          return {
            scrollTop: n,
            scrollHeight: o,
            jump: 0,
            changed: !1
          };
        },
        { scrollHeight: 0, jump: 0, scrollTop: 0, changed: !1 }
      ),
      v((t) => t.changed),
      g((t) => t.jump)
    ),
    Pn
  );
}), $t = c(Pt, (e) => {
  e.link(
    e.pipe(
      w,
      ie(
        (t, n) => {
          if (n < 0)
            return { direction: Kt, prevScrollTop: 0 };
          if (e.getValue(qt))
            return { direction: t.direction, prevScrollTop: n };
          const i = n === t.prevScrollTop && n === 0;
          return {
            direction: n < t.prevScrollTop || i ? Kt : Pt,
            prevScrollTop: n
          };
        },
        { direction: Pt, prevScrollTop: 0 }
      ),
      g((t) => t.direction)
    ),
    $t
  ), e.link(e.pipe(w, xe(100), P(Ko)), $t);
}), vn = c(0, (e) => {
  e.link(e.pipe(Xe, v(Yo), P(0)), vn), e.link(
    e.pipe(
      w,
      nn(100),
      k(Xe),
      v(([, t]) => !!t),
      ie(([, t], [n]) => [t, n], [0, 0]),
      g(([t, n]) => n - t)
    ),
    vn
  );
});
function rn(e, t) {
  if (typeof e == "number")
    return {
      index: e,
      offset: 0,
      behavior: "auto",
      align: "start-no-overflow"
    };
  const n = {
    index: Number.NaN,
    align: e.align ?? "start-no-overflow",
    behavior: e.behavior ?? "auto",
    offset: e.offset ?? 0
  };
  return e.index === "LAST" ? n.index = t : e.index < 0 ? n.index = t + e.index : n.index = e.index, n;
}
function Wn({
  location: e,
  sizeTree: t,
  offsetTree: n,
  totalHeight: o,
  totalCount: i,
  viewportHeight: s,
  headerHeight: l,
  stickyHeaderHeight: r,
  stickyFooterHeight: a
}) {
  const { align: u, behavior: p, offset: d, index: I } = rn(e, i - 1);
  function b() {
    const L = he(t, I)[1];
    if (L === void 0)
      throw new Error(`Item at index ${I} not found`);
    return L;
  }
  s -= r + a;
  let f = Je(I, n) + l - r;
  u === "end" ? f = f - s + b() : u === "center" && (f = f - s / 2 + b() / 2), d && (f += d);
  let m = 0;
  return u === "start" && (m = Math.max(0, Math.min(f - (o - s)))), f = Math.max(0, f), { top: f, behavior: p, align: u, forceBottomSpace: m };
}
const Ke = c(null), Xo = c(!1), Ue = c(!0), Ut = E((e) => {
  e.link(
    e.pipe(
      Ut,
      g(() => !0)
    ),
    Ue
  ), e.link(
    e.pipe(
      Ut,
      g(() => null)
    ),
    Ke
  );
}), zn = E((e) => {
  e.link(
    e.pipe(
      zn,
      k(Ce, ge, rt),
      g(([t, n, o, i]) => {
        let { align: s, behavior: l, offset: r, index: a } = rn(t, n - 1);
        const u = typeof t != "number" ? t.done : void 0, [p, d] = Nn(a, o);
        return p < -i.listOffset ? ((typeof t == "number" || t.align === void 0) && (s = "start-no-overflow"), { index: a, align: s, behavior: l, offset: r, done: u }) : p + d > -i.listOffset + i.visibleListHeight ? ((typeof t == "number" || t.align === void 0) && (s = "end"), { index: a, align: s, behavior: l, offset: r, done: u }) : null;
      }),
      v((t) => t !== null)
    ),
    // @ts-expect-error contra variance
    Z
  );
}), Z = E((e) => {
  const t = e.pipe(
    Z,
    k(
      Q,
      ge,
      Ce,
      K,
      lt,
      ze,
      st,
      We
    ),
    g(
      ([n, o, i, s, l, r, a, u, p]) => {
        try {
          return Wn({
            location: n,
            totalHeight: p,
            sizeTree: o,
            offsetTree: i,
            totalCount: s,
            viewportHeight: l,
            headerHeight: r,
            stickyHeaderHeight: a,
            stickyFooterHeight: u
          });
        } catch {
          return null;
        }
      }
    ),
    v((n) => n !== null)
  );
  e.link(Z, Ke), e.link(t, $e), e.link(
    e.pipe(
      Z,
      v((n) => typeof n != "number" && n.index === "LAST"),
      P(!0)
    ),
    Se
  ), e.link(e.pipe(t, P(!1)), Ue), e.link(e.pipe(t, P(!1)), Xo), e.link(
    e.pipe(
      Q,
      // wait for the list to render with the specified sizeTree, so that enough space is available to scroll by
      xe(0),
      k(Ue, Ke),
      v(([, n, o]) => !n && o !== null),
      g(([, , n]) => n)
    ),
    Z
  ), e.sub(e.pipe(ot, xe(10)), () => {
    const n = e.getValue(Ke);
    n !== null && typeof n != "number" && n.done !== void 0 && n.done(), e.pubIn({
      [Ke]: null,
      [Ue]: !0
    });
  }), e.link(
    e.pipe(
      Tt,
      // wait for the list to render with the specified scrollOffset, so that enough space is available to scroll by
      Ze(),
      v((n) => n !== 0)
    ),
    X
  ), e.link(
    e.pipe(
      Tt,
      xn(w),
      g(() => 0)
    ),
    Tt
  );
}), ke = c(null), Fe = c(
  null,
  (e) => {
    e.link(
      e.pipe(
        Fe,
        v((n) => n !== null)
      ),
      ke
    );
    const t = e.pipe(
      e.combine(Fe, Q),
      k(ke),
      v(([[n, o], i]) => n !== null && !T(o) && i !== null),
      g(([[n]]) => n)
    );
    e.link(e.pipe(t, Ze()), Z), e.link(
      e.pipe(
        t,
        xn(
          e.pipe(
            Ue,
            v((n) => n)
          )
        ),
        P(null)
        // unset the location after the scroll completes
      ),
      ke
    );
  },
  !1
);
function Qo(e, t) {
  var l, r;
  const n = t.slice();
  let o = 0;
  const i = [];
  for (const { k: a, v: u } of de(e)) {
    for (; n.length && n[0] < a; )
      n.shift(), o++;
    const p = Math.max(0, a - o), d = ((l = i.at(-1)) == null ? void 0 : l.k) ?? -1;
    p === d ? (((r = i.at(-2)) == null ? void 0 : r.v) ?? -1) === u ? i.pop() : i[i.length - 1].v = u : i.push({ k: p, v: u });
  }
  let s = Re();
  for (const { k: a, v: u } of i)
    s = H(s, a, u);
  return s;
}
const Ce = c(0), He = c(null), S = c(null, (e) => {
  e.link(
    e.pipe(
      S,
      v((t) => t !== null),
      g((t) => t.length)
    ),
    Ce
  );
}), je = c(null), oe = E((e) => {
  e.link(
    e.pipe(
      oe,
      k(Ee),
      g(([n, o]) => -(o * n.length))
    ),
    Y
  ), e.link(e.pipe(oe, P(!0)), Ge), e.link(e.pipe(oe, Ze()), je);
  function t(n, o) {
    e.pubIn({
      [St]: !0,
      [X]: n,
      [Et]: n
    }), o ? requestAnimationFrame(() => {
      e.pubIn({
        [Y]: 0,
        [Et]: 0,
        [je]: null,
        [Ge]: !1,
        [St]: !1
      });
    }) : e.pubIn({
      [Y]: 0,
      [Et]: 0,
      [je]: null,
      [Ge]: !1,
      [St]: !1
    });
  }
  e.sub(
    e.pipe(
      ge,
      k(je),
      v(([, n]) => n !== null),
      g(([n, o]) => {
        if (o === null)
          throw new Error("Unexpected null items");
        return Je(o.length, n);
      })
    ),
    (n) => {
      t(n, !1);
    }
  ), e.sub(
    e.pipe(
      oe,
      Ct(2),
      k(ge, je),
      v(([, , n]) => n !== null),
      g(([n, o]) => Je(n.length, o))
    ),
    (n) => {
      t(n, !0);
    }
  ), e.changeWith(S, oe, (n, o) => n ? [...o, ...n] : o.slice()), e.link(
    e.pipe(
      oe,
      k(Q, Ee),
      g(([n, o, i]) => {
        const s = n.length, l = i;
        return de(o).reduce(
          (a, { k: u, v: p }) => ({
            ranges: [...a.ranges, { startIndex: a.prevIndex, endIndex: u + s - 1, size: a.prevSize }],
            prevIndex: u + s,
            prevSize: p
          }),
          {
            ranges: [],
            prevIndex: 0,
            prevSize: l
          }
        ).ranges;
      })
    ),
    fe
  );
}), nt = E((e) => {
  const t = e.pipe(
    nt,
    k(rt, un, He, Q),
    v(([, , , , o]) => !T(o)),
    g(([{ data: o, scrollToBottom: i }, s, l, r]) => {
      if (i === !1 || i === void 0)
        return null;
      let a = "auto";
      const u = s.isAtBottom;
      if (typeof i == "function") {
        const p = i({ data: o, scrollLocation: s, scrollInProgress: l, context: r, atBottom: u });
        if (!p)
          return null;
        if (typeof p == "object")
          return p;
        if (typeof p == "number")
          return { index: p, align: "end", behavior: "auto" };
        a = p;
      } else {
        if (!u)
          return null;
        a = i;
      }
      return a === !0 && (a = "auto"), { index: "LAST", align: "end", behavior: a };
    })
  );
  e.link(
    e.pipe(
      t,
      v((o) => o !== null),
      g(() => !0)
    ),
    Se
  ), e.link(
    e.pipe(
      ot,
      k(Se),
      v(([o, i]) => i),
      g(() => !1)
    ),
    Se
  );
  const n = e.pipe(
    pn,
    k(Se),
    v(([o, i]) => o === "up" && i)
  );
  e.link(
    e.pipe(
      n,
      g(() => !1)
    ),
    Se
  ), e.link(e.pipe(n, P(!0)), Ut), e.link(
    e.pipe(
      t,
      v((o) => o !== null),
      xe(20)
    ),
    Z
  );
}), wt = E((e) => {
  e.changeWith(S, wt, (t, n) => t ? [...t, ...n.data] : n.data.slice()), e.link(wt, nt);
}), It = E((e) => {
  e.changeWith(S, It, (t, n) => t ? [...t.slice(0, n.offset), ...n.data, ...t.slice(n.offset)] : n.data.slice()), e.changeWith(J, It, ([t], n) => {
    const i = he(t, n.offset, "k")[0], s = n.data.length;
    return [Ao(t, i, Number.POSITIVE_INFINITY, (r, a) => [r + s, a]), i];
  }), e.link(It, nt);
}), Gt = E((e) => {
  e.changeWith(S, Gt, (t, { offset: n, count: o }) => t ? t.slice(0, n).concat(t.slice(n + o)) : []), e.changeWith(J, Gt, ([t], { offset: n, count: o }) => [Oo(t, n, o), n]);
}), Wt = c(null), Ne = E((e) => {
  e.sub(
    e.pipe(
      Ne,
      k(S),
      v(([{ purgeItemSizes: t }, n]) => !!t || n === null || n.length === 0)
    ),
    ([t, n]) => {
      n === null || n.length === 0 ? e.pubIn({
        ...t.initialLocation ? { [Fe]: t.initialLocation } : {},
        [S]: t.data.slice()
      }) : e.pubIn({
        ...t.initialLocation ? { [Fe]: t.initialLocation } : {},
        [J]: tt,
        [Me]: De,
        [Wt]: t.data.slice()
      });
    }
  ), e.sub(
    e.pipe(
      Xn,
      k(Wt),
      Ct(),
      v(([, t]) => t !== null)
    ),
    ([, t]) => {
      e.pubIn({
        [S]: t,
        [Wt]: null
      });
    }
  ), e.link(
    e.pipe(
      Ne,
      v(({ purgeItemSizes: t }) => !t),
      k(Ee),
      v(([, t]) => t > 0),
      g(([{ data: t }, n]) => [
        {
          size: n,
          startIndex: t.length,
          endIndex: Number.POSITIVE_INFINITY
        }
      ])
    ),
    fe
  ), e.sub(
    e.pipe(
      Ne,
      v(({ purgeItemSizes: t }) => !t)
    ),
    ({ data: t, initialLocation: n, suppressItemMeasure: o }) => {
      requestAnimationFrame(() => {
        o || e.pub(Jn), requestAnimationFrame(() => {
          n && e.pubIn({
            [Z]: n
          });
        });
      }), e.pubIn({
        [S]: t.slice()
      });
    }
  );
}), vt = E((e) => {
  e.link(
    e.pipe(
      vt,
      k(ge),
      g(([t, n]) => -Je(t, n))
    ),
    X
  ), e.changeWith(S, e.pipe(vt, Ze()), (t, n) => t ? t.slice(n) : []), e.changeWith(J, e.pipe(vt, Ze()), ([t], n) => [de(t).reduce((i, { k: s, v: l }) => H(i, Math.max(0, s - n), l), Re()), 0]);
}), Sn = E((e) => {
  e.changeWith(S, Sn, (t, n) => t ? t.slice(0, t.length - n) : []), e.link(
    e.pipe(
      Sn,
      k(Ce, Ee),
      g(([, t, n]) => [
        {
          size: n,
          startIndex: t,
          endIndex: Number.POSITIVE_INFINITY
        }
      ])
    ),
    fe
  );
}), Yn = E((e) => {
  const t = e.pipe(
    Yn,
    k(S),
    g(([n, o]) => {
      if (!o) return [];
      const i = [];
      return o.forEach((s, l) => {
        n(s, l) && i.push(l);
      }), i;
    })
  );
  e.changeWith(S, t, (n, o) => n ? n.filter((i, s) => !o.includes(s)) : []), e.changeWith(J, t, ([n], o) => [Qo(n, o), 0]);
}), Zt = E((e) => {
  e.changeWith(S, Zt, (t, { mapper: n }) => t ? t.map(n) : []), e.link(
    e.pipe(
      Zt,
      Ct(3),
      k(ln),
      v(([{ autoscrollToBottomBehavior: t }, n]) => n && !!t),
      g(([{ autoscrollToBottomBehavior: t }]) => typeof t == "object" ? t.location() : { index: "LAST", align: "end", behavior: t }),
      v((t) => !!t)
    ),
    Z
  );
}), Jt = E((e) => {
  e.changeWith(S, Jt, (t, { newData: n }) => n), e.link(
    e.pipe(
      Jt,
      Ct(3),
      k(ln),
      v(([{ autoscrollToBottomBehavior: t }, n]) => n && !!t),
      g(([{ autoscrollToBottomBehavior: t }]) => typeof t == "object" ? t.location() : { index: "LAST", align: "end", behavior: t }),
      v((t) => !!t)
    ),
    Z
  );
});
function ei(e, t) {
  return [
    {
      data: t == null ? void 0 : t[e],
      prevData: (t == null ? void 0 : t[e - 1]) ?? null,
      nextData: (t == null ? void 0 : t[e + 1]) ?? null,
      height: 0,
      index: e,
      offset: 0,
      type: "flat"
    }
  ];
}
const ti = [], De = {
  items: ti,
  listBottom: 0,
  listTop: 0,
  offsetTree: [],
  paddingBottom: 0,
  paddingTop: 0,
  totalCount: 0,
  totalHeight: 0,
  deviationDelta: 0,
  data: null
}, St = c(!1), Me = c(De, (e) => {
  e.link(
    e.pipe(
      e.combine(
        ni,
        Un,
        Q,
        ge,
        Ce,
        We,
        S,
        Tt,
        Fe,
        ke,
        it,
        ze,
        st,
        Y,
        St,
        Te,
        G
      ),
      v((t) => {
        const n = t.at(-1), o = t.at(-2), i = t.at(-3);
        return !o && !n && !i;
      }),
      ie(
        (t, [
          n,
          o,
          i,
          s,
          l,
          r,
          a,
          u,
          p,
          d,
          I,
          b,
          f,
          m
        ]) => {
          var ct;
          if ((a == null ? void 0 : a.length) === 0)
            return De;
          if (T(i)) {
            let ne = 0;
            return p !== null && (ne = rn(p, l - 1).index), { ...De, items: ei(ne, a), offsetTree: s, totalCount: l, data: a };
          }
          let L = 0;
          d !== null && n === 0 && (L = Wn({
            totalHeight: r,
            location: d,
            sizeTree: i,
            offsetTree: s,
            totalCount: l,
            viewportHeight: e.getValue(K),
            headerHeight: e.getValue(lt),
            stickyHeaderHeight: b,
            stickyFooterHeight: f
          }).top ?? 0);
          let q = 0;
          e.getValue(w) !== 0 && !e.getValue(Mt) && e.getValue($t) === Kt && t.totalCount === l && t.items.length > 0 && (q = r - t.totalHeight, q !== 0 && (q += e.getValue(Pn)));
          const U = e.getValue(Kn), j = Math.min(
            Math.max(
              n + L + u - m - I + q - U,
              0
            ),
            r - o
          ), be = j + o + U * 2;
          if (t.offsetTree === s && t.totalCount === l && t.data === a && j >= t.listTop && be <= t.listBottom)
            return t;
          const ee = [], ut = l - 1, se = 0, W = Fo(s, j, be, se);
          let C = 0, O = 0, R = !1;
          for (const ne of W) {
            const {
              value: { offset: Ie, height: le }
            } = ne;
            let re = ne.start;
            C = Ie, Ie < j && (re += Math.floor((j - Ie) / le), C += (re - ne.start) * le), re < se && (C += (se - re) * le, re = se);
            const Dt = Math.min(ne.end, ut);
            for (let ue = re; ue <= Dt && !(C >= be); ue++) {
              const Nt = {
                data: a == null ? void 0 : a[ue],
                prevData: (a == null ? void 0 : a[ue - 1]) ?? null,
                nextData: (a == null ? void 0 : a[ue + 1]) ?? null,
                height: le,
                index: ue,
                offset: C,
                type: "flat"
              };
              R || (R = !0, O = C), ee.push(Nt), C += le;
            }
          }
          ee.length === 0 && (O = C = 0);
          const te = r - C, Ve = ((ct = ee[0]) == null ? void 0 : ct.offset) || 0;
          return { items: ee, listBottom: C, listTop: O, offsetTree: s, paddingBottom: te, paddingTop: Ve, totalCount: l, totalHeight: r, data: a, deviationDelta: q };
        },
        De
      )
    ),
    Me
  );
}), Lt = Le([], (e) => e.pipe(
  e.combine(Me, w),
  g(([t, n]) => {
    const o = t.items.slice();
    for (; o.length > 0 && o[0].offset + o[0].height < n; )
      o.shift();
    return o.map((i) => i.data);
  })
)), G = c(!1), Ge = c(!1), kt = E((e) => {
  e.link(
    e.pipe(
      Me,
      g((t) => t.deviationDelta),
      v((t) => t !== 0)
    ),
    kt
  ), _n() ? (e.sub(e.pipe(kt, k(Y, w)), ([t, n]) => {
    e.pub(Y, n - t);
  }), e.sub(
    e.pipe(e.combine(w, Y, Te, Ge)),
    ([t, n, o, i]) => {
      o || i || (n > 0 && t < n ? (e.pub(G, !0), e.pub($e, { top: 0, behavior: "instant" }), setTimeout(() => {
        e.pubIn({
          [G]: !1,
          [Y]: 0
        });
      })) : n < 0 && t <= 0 && (e.pubIn({
        [G]: !0,
        [Y]: 0
      }), setTimeout(() => {
        e.pub($e, { top: 0, behavior: "instant" }), e.pub(G, !1);
      })));
    }
  ), e.sub(
    e.pipe(
      e.combine(Xe, Y, G, Te, Ge),
      v(
        ([t, n, o, i, s]) => !t && n !== 0 && !o && !i && !s
      ),
      nn(100)
    ),
    ([, t]) => {
      e.pub(G, !0), t < 0 ? requestAnimationFrame(() => {
        e.pub(X, -t), e.pub(Y, 0), requestAnimationFrame(() => {
          e.pub(G, !1);
        });
      }) : requestAnimationFrame(() => {
        e.pub(X, -t), e.pub(Y, 0), requestAnimationFrame(() => {
          e.pub(G, !1);
        });
      });
    }
  )) : e.link(kt, X);
}), Mt = E(), un = c(!1), ot = E((e) => {
  e.link(e.pipe(ot, P(!1)), Mt);
}, !1), w = c(0), K = c(0), xt = c(0), D = c(0), ni = w, Tt = c(0), ze = c(0), it = c(0), st = c(0), cn = c(0), ye = c(null), jn = on(), Kn = c(0), oi = jo, ii = 50, lt = Le(0, (e) => e.pipe(
  e.combine(ze, it),
  g(([t, n]) => t + n)
)), qn = Le(0, (e) => e.pipe(
  e.combine(st, cn),
  g(([t, n]) => t + n)
)), si = Le(0, (e) => e.pipe(
  e.combine(ze, it, w),
  g(([t, n, o]) => t + Math.max(n - o, 0))
)), li = Le(0, (e) => e.pipe(
  e.combine(st, cn, w, K, D),
  g(([t, n, o, i, s]) => {
    o = Math.min(o, s - i);
    const l = Math.max(n - (s - (o + i)), 0);
    return t + l;
  })
)), Un = Le(0, (e) => e.pipe(
  e.combine(K, si, li),
  g(([t, n, o]) => Math.max(0, t - n - o))
)), Vt = c(0), Gn = c(0, (e) => {
  e.link(
    e.pipe(
      e.combine(Gn, We, K, lt, ze),
      g(([t, n, o, i, s]) => t === 0 ? 0 : Math.max(0, Math.min(t - (n + i + s - o))))
    ),
    Vt
  );
}), $e = E((e) => {
  e.link(
    e.pipe(
      $e,
      g((t) => t.align === "start" ? t.top ?? 0 : 0)
    ),
    Gn
  ), e.link(
    e.pipe(
      $e,
      k(w),
      v(([t, n]) => t.top !== n),
      P(!0)
    ),
    Mt
  );
}), an = E((e) => {
  e.link(
    e.pipe(
      $e,
      k(et),
      g(([t, n]) => ("top" in t && typeof t.top < "u" && (t = { ...t, top: t.top + n }), t))
    ),
    an
  );
}), rt = Le(
  {
    listOffset: 0,
    visibleListHeight: 0,
    scrollHeight: 0,
    bottomOffset: 0,
    isAtBottom: !1
  },
  (e) => e.pipe(
    e.combine(
      w,
      lt,
      qn,
      it,
      Un,
      D,
      Vt,
      Te,
      ke,
      G,
      Se
    ),
    v(([, , , , , , , t, n, o]) => !t && n === null && !o),
    g(
      ([
        t,
        n,
        o,
        i,
        s,
        l,
        r,
        a,
        u,
        p,
        d
      ]) => {
        const I = e.getValue(Hn), b = l - n - o, f = -t + i, m = b + Math.min(0, f) - s - r;
        return {
          scrollHeight: b,
          listOffset: f,
          visibleListHeight: s,
          bottomOffset: m,
          isAtBottom: d || m <= I
        };
      }
    )
  )
), Xt = E((e) => {
  e.link(
    e.pipe(
      w,
      xe(0),
      k(rt, ke, Te),
      v(([, t, n, o]) => t.scrollHeight > 0 && n == null && !o),
      g(([, t]) => t)
    ),
    Xt
  );
}), X = E(), Y = c(0), Et = c(0), Rt = c(0), Zn = c(""), pn = E(), Jn = on(), Xn = on(), fn = c(!1), Qe = c(null);
c(0);
const et = c(0, (e) => {
  e.link(
    e.pipe(
      e.combine(Pe, me, et),
      g(([t, n, o]) => t - Math.max(0, o - n))
    ),
    K
  );
}), me = c(0, (e) => {
  e.link(
    e.pipe(
      e.combine(me, et),
      g(([t, n]) => Math.max(0, t - n))
    ),
    w
  );
}), Pe = c(0);
function Qn(e) {
  return {
    data: {
      prepend: (t) => {
        e.pub(oe, t);
      },
      append: (t, n) => {
        e.pub(wt, {
          data: t,
          scrollToBottom: n
        });
      },
      replace: (t, n) => {
        e.pub(Ne, {
          ...n,
          data: t
        });
      },
      map: (t, n) => {
        e.pub(Zt, {
          mapper: t,
          autoscrollToBottomBehavior: n
        });
      },
      findAndDelete: (t) => {
        e.pub(Yn, t);
      },
      findIndex: (t) => e.getValue(S).findIndex(t),
      find: (t) => e.getValue(S).find(t),
      insert: (t, n, o) => {
        e.pub(It, {
          data: t,
          offset: n,
          scrollToBottom: o
        });
      },
      deleteRange: (t, n) => {
        e.pub(Gt, {
          offset: t,
          count: n
        });
      },
      batch: (t, n) => {
        e.pub(Te, !0), t(), e.pub(Te, !1), e.pub(nt, { data: [], scrollToBottom: n });
      },
      get: () => e.getValue(S).slice(),
      getCurrentlyRendered: () => e.getValue(Lt),
      removeFromStart: (t) => {
        e.pub(vt, t);
      }
    },
    scrollToItem: (t) => {
      e.pub(Z, t);
    },
    scrollIntoView: (t) => {
      e.pub(zn, t);
    },
    scrollerElement: () => e.getValue(ye),
    getScrollLocation() {
      return e.getValue(rt);
    },
    cancelSmoothScroll() {
      e.pub(jn);
    },
    height: (t) => {
      var i;
      const n = ((i = e.getValue(S)) == null ? void 0 : i.indexOf(t)) ?? -1;
      if (n === -1)
        return 0;
      const o = e.getValue(Q);
      return he(o, n)[1] ?? 0;
    }
  };
}
function Fi() {
  return M(rt);
}
function Hi() {
  return M(Lt);
}
function Pi() {
  const e = we();
  return h.useMemo(() => Qn(e), [e]);
}
const Wi = {
  prepend: "prepend",
  removeFromStart: "remove-from-start",
  removeFromEnd: "remove-from-end"
}, eo = c(null), to = c(null), no = c(null), oo = c(null), io = c(null), so = c("div"), ri = {
  position: "sticky",
  top: 0,
  zIndex: 1
}, dt = {
  overflowAnchor: "none"
}, ui = {
  position: "sticky",
  bottom: 0
}, lo = h.forwardRef((e, t) => /* @__PURE__ */ y("div", { style: { zIndex: 1 }, ...e, ref: t })), ro = h.forwardRef((e, t) => /* @__PURE__ */ y("div", { ...e, ref: t })), uo = h.forwardRef(
  ({ style: e, ...t }, n) => /* @__PURE__ */ y("div", { ...t, style: { ...ri, ...e }, ref: n })
), co = h.forwardRef(
  ({ style: e, ...t }, n) => /* @__PURE__ */ y("div", { ...t, style: { ...ui, ...e }, ref: n })
), ao = c(lo), po = c(uo), fo = c(ro), ho = c(co), go = ({ index: e }) => /* @__PURE__ */ En("div", { children: [
  "Item ",
  e
] }), mo = ({ index: e }) => e, Qt = c(go), bo = c(mo), en = (e) => e, Io = c(en), vo = c(
  null,
  (e) => {
    e.sub(
      e.pipe(vo, k(S, Io, ge, J)),
      ([t, n, o, i, s]) => {
        if (t === void 0)
          return;
        if (!t || !t.data || !t.data.length) {
          e.pubIn({
            [S]: [],
            [J]: tt,
            [Me]: De
          });
          return;
        }
        const l = t.data, r = t.scrollModifier;
        if (r === "prepend") {
          if (n === null || !n.length) {
            e.pub(S, l);
            return;
          }
          const a = n[0], u = l.findIndex((I) => o(I) === o(a)), p = u === -1 ? l : l.slice(0, u), d = u === -1 ? [] : l.slice(u);
          e.pubIn({
            [S]: d
          }), e.pubIn({
            [oe]: p
          });
          return;
        }
        if (r === "remove-from-start") {
          const a = l[0], u = (n == null ? void 0 : n.findIndex((d) => o(d) === o(a))) ?? -1;
          if (u === -1) {
            e.pub(S, l);
            return;
          }
          const p = Je(u, i);
          e.pub(X, -p), queueMicrotask(() => {
            e.pub(S, l);
            const d = de(s[0]).reduce((I, { k: b, v: f }) => H(I, Math.max(0, b - u), f), Re());
            e.pub(J, [d, 0]);
          });
          return;
        }
        if (r === "remove-from-end") {
          e.pub(S, l), e.pub(fe, [
            {
              size: e.getValue(Ee),
              startIndex: l.length,
              endIndex: Number.POSITIVE_INFINITY
            }
          ]);
          return;
        }
        if ((r == null ? void 0 : r.type) === "item-location") {
          l !== n && e.pubIn({
            [Ne]: {
              data: l,
              initialLocation: r.location,
              purgeItemSizes: r.purgeItemSizes
            }
          });
          return;
        }
        if ((r == null ? void 0 : r.type) === "auto-scroll-to-bottom") {
          e.pubIn({
            [S]: l,
            [nt]: {
              data: l,
              scrollToBottom: r.autoScroll
            }
          });
          return;
        }
        if ((r == null ? void 0 : r.type) === "items-change") {
          e.pub(Jt, {
            newData: l,
            autoscrollToBottomBehavior: r.behavior
          });
          return;
        }
        e.pub(S, l);
      }
    );
  },
  (e, t) => e ? e.data === (t == null ? void 0 : t.data) : !1
), ci = ({ item: e, ItemContent: t, mount: n, unmount: o }) => {
  const i = M(He), s = h.useRef(null), l = h.useCallback(
    (r) => {
      r ? (s.current = r, n(r)) : s.current && (o(s.current), s.current = null);
    },
    [n, o]
  );
  return /* @__PURE__ */ y(
    "div",
    {
      ref: l,
      "data-index": e.index,
      "data-known-size": e.height,
      style: {
        overflowAnchor: "none",
        position: "absolute",
        width: "100%",
        top: e.offset
      },
      children: /* @__PURE__ */ y(t, { index: e.index, prevData: e.prevData, nextData: e.nextData, data: e.data, context: i })
    }
  );
}, tn = c("top", (e) => {
  e.link(
    e.pipe(
      e.combine(tn, We, K, lt, qn),
      v(([t]) => t === "bottom" || t === "bottom-smooth"),
      g(([, t, n, o, i]) => Math.max(0, n - t - o - i))
    ),
    Rt
  ), e.link(
    e.pipe(
      e.combine(Rt, tn),
      v(([, t]) => t === "bottom-smooth"),
      ie(
        (t, [n]) => [t[1], n],
        [0, 0]
      ),
      g(([t, n]) => t > 0 && n > 0 ? "margin-top 0.2s ease-out" : "")
    ),
    Zn
  );
});
function gt(e) {
  const t = h.useRef(null);
  return [h.useCallback(
    (o) => {
      o ? (t.current = o, e == null || e.observe(o, { box: "border-box" })) : t.current && (e == null || e.unobserve(t.current), t.current = null);
    },
    [e]
  ), t];
}
function ai(e, t) {
  return Math.abs(e - t) < 0.5;
}
function pi(e, t, n) {
  const o = we(), i = h.useRef(null), s = h.useRef(null), l = h.useCallback(() => {
    i.current && (cancelAnimationFrame(i.current), i.current = null, s.current = null);
  }, []);
  h.useEffect(() => o.sub(pn, (u) => {
    u !== s.current && l();
  }), [o, l]), h.useEffect(() => o.sub(jn, l), [o, l]);
  const r = h.useCallback(
    (u, p, d) => {
      var L;
      i.current && l();
      const I = ((L = e.current) == null ? void 0 : L.scrollTop) ?? 0;
      s.current = I < u ? "down" : "up";
      let b = 0, f = 0;
      function m() {
        var U, j;
        const q = I + (u - I) * p(b);
        (U = e.current) == null || U.scrollTo({ top: q, behavior: "instant" }), b += 1 / d, f += 1, f < d ? i.current = requestAnimationFrame(m) : ((j = e.current) == null || j.scrollTo({ top: u, behavior: "instant" }), i.current = null, s.current = null);
      }
      m();
    },
    [e, l]
  );
  return h.useCallback(
    (u) => {
      var d, I;
      const p = e.current;
      if (!(!p || u.top === void 0)) {
        if (ai(u.top, p.scrollTop)) {
          requestAnimationFrame(() => {
            var b;
            o.pub(ot, (b = e.current) == null ? void 0 : b.scrollTop);
          });
          return;
        }
        if (n.current = u.top, o.pub(un, !0), u.forceBottomSpace !== void 0 && t.current && (t.current.style.paddingBottom = `${u.forceBottomSpace}px`), u.behavior === "smooth")
          r(u.top ?? 0, oi, ii);
        else if (u.behavior === "auto" || u.behavior === "instant" || u.behavior === void 0)
          l(), (d = e.current) == null || d.scrollTo(u);
        else {
          const { easing: b, animationFrameCount: f } = u.behavior(((I = e.current) == null ? void 0 : I.scrollTop) ?? 0, u.top ?? 0);
          r(u.top ?? 0, b, f);
        }
      }
    },
    [o, r, t, e, n, l]
  );
}
function fi(e) {
  return hi(gi(mi(di(e), 8 * e.length))).toLowerCase();
}
function hi(e) {
  for (var t, n = "0123456789ABCDEF", o = "", i = 0; i < e.length; i++)
    t = e.charCodeAt(i), o += n.charAt(t >>> 4 & 15) + n.charAt(15 & t);
  return o;
}
function di(e) {
  for (var t = Array(e.length >> 2), n = 0; n < t.length; n++) t[n] = 0;
  for (n = 0; n < 8 * e.length; n += 8) t[n >> 5] |= (255 & e.charCodeAt(n / 8)) << n % 32;
  return t;
}
function gi(e) {
  for (var t = "", n = 0; n < 32 * e.length; n += 8) t += String.fromCharCode(e[n >> 5] >>> n % 32 & 255);
  return t;
}
function mi(e, t) {
  e[t >> 5] |= 128 << t % 32, e[14 + (t + 64 >>> 9 << 4)] = t;
  for (var n = 1732584193, o = -271733879, i = -1732584194, s = 271733878, l = 0; l < e.length; l += 16) {
    const r = n, a = o, u = i, p = s;
    o = F(
      o = F(
        o = F(
          o = F(
            o = _(
              o = _(
                o = _(
                  o = _(
                    o = B(
                      o = B(
                        o = B(
                          o = B(
                            o = N(
                              o = N(
                                o = N(
                                  o = N(
                                    o,
                                    i = N(
                                      i,
                                      s = N(s, n = N(n, o, i, s, e[l + 0], 7, -680876936), o, i, e[l + 1], 12, -389564586),
                                      n,
                                      o,
                                      e[l + 2],
                                      17,
                                      606105819
                                    ),
                                    s,
                                    n,
                                    e[l + 3],
                                    22,
                                    -1044525330
                                  ),
                                  i = N(
                                    i,
                                    s = N(s, n = N(n, o, i, s, e[l + 4], 7, -176418897), o, i, e[l + 5], 12, 1200080426),
                                    n,
                                    o,
                                    e[l + 6],
                                    17,
                                    -1473231341
                                  ),
                                  s,
                                  n,
                                  e[l + 7],
                                  22,
                                  -45705983
                                ),
                                i = N(
                                  i,
                                  s = N(s, n = N(n, o, i, s, e[l + 8], 7, 1770035416), o, i, e[l + 9], 12, -1958414417),
                                  n,
                                  o,
                                  e[l + 10],
                                  17,
                                  -42063
                                ),
                                s,
                                n,
                                e[l + 11],
                                22,
                                -1990404162
                              ),
                              i = N(
                                i,
                                s = N(s, n = N(n, o, i, s, e[l + 12], 7, 1804603682), o, i, e[l + 13], 12, -40341101),
                                n,
                                o,
                                e[l + 14],
                                17,
                                -1502002290
                              ),
                              s,
                              n,
                              e[l + 15],
                              22,
                              1236535329
                            ),
                            i = B(
                              i,
                              s = B(s, n = B(n, o, i, s, e[l + 1], 5, -165796510), o, i, e[l + 6], 9, -1069501632),
                              n,
                              o,
                              e[l + 11],
                              14,
                              643717713
                            ),
                            s,
                            n,
                            e[l + 0],
                            20,
                            -373897302
                          ),
                          i = B(
                            i,
                            s = B(s, n = B(n, o, i, s, e[l + 5], 5, -701558691), o, i, e[l + 10], 9, 38016083),
                            n,
                            o,
                            e[l + 15],
                            14,
                            -660478335
                          ),
                          s,
                          n,
                          e[l + 4],
                          20,
                          -405537848
                        ),
                        i = B(
                          i,
                          s = B(s, n = B(n, o, i, s, e[l + 9], 5, 568446438), o, i, e[l + 14], 9, -1019803690),
                          n,
                          o,
                          e[l + 3],
                          14,
                          -187363961
                        ),
                        s,
                        n,
                        e[l + 8],
                        20,
                        1163531501
                      ),
                      i = B(
                        i,
                        s = B(s, n = B(n, o, i, s, e[l + 13], 5, -1444681467), o, i, e[l + 2], 9, -51403784),
                        n,
                        o,
                        e[l + 7],
                        14,
                        1735328473
                      ),
                      s,
                      n,
                      e[l + 12],
                      20,
                      -1926607734
                    ),
                    i = _(
                      i,
                      s = _(s, n = _(n, o, i, s, e[l + 5], 4, -378558), o, i, e[l + 8], 11, -2022574463),
                      n,
                      o,
                      e[l + 11],
                      16,
                      1839030562
                    ),
                    s,
                    n,
                    e[l + 14],
                    23,
                    -35309556
                  ),
                  i = _(
                    i,
                    s = _(s, n = _(n, o, i, s, e[l + 1], 4, -1530992060), o, i, e[l + 4], 11, 1272893353),
                    n,
                    o,
                    e[l + 7],
                    16,
                    -155497632
                  ),
                  s,
                  n,
                  e[l + 10],
                  23,
                  -1094730640
                ),
                i = _(
                  i,
                  s = _(s, n = _(n, o, i, s, e[l + 13], 4, 681279174), o, i, e[l + 0], 11, -358537222),
                  n,
                  o,
                  e[l + 3],
                  16,
                  -722521979
                ),
                s,
                n,
                e[l + 6],
                23,
                76029189
              ),
              i = _(
                i,
                s = _(s, n = _(n, o, i, s, e[l + 9], 4, -640364487), o, i, e[l + 12], 11, -421815835),
                n,
                o,
                e[l + 15],
                16,
                530742520
              ),
              s,
              n,
              e[l + 2],
              23,
              -995338651
            ),
            i = F(
              i,
              s = F(s, n = F(n, o, i, s, e[l + 0], 6, -198630844), o, i, e[l + 7], 10, 1126891415),
              n,
              o,
              e[l + 14],
              15,
              -1416354905
            ),
            s,
            n,
            e[l + 5],
            21,
            -57434055
          ),
          i = F(
            i,
            s = F(s, n = F(n, o, i, s, e[l + 12], 6, 1700485571), o, i, e[l + 3], 10, -1894986606),
            n,
            o,
            e[l + 10],
            15,
            -1051523
          ),
          s,
          n,
          e[l + 1],
          21,
          -2054922799
        ),
        i = F(
          i,
          s = F(s, n = F(n, o, i, s, e[l + 8], 6, 1873313359), o, i, e[l + 15], 10, -30611744),
          n,
          o,
          e[l + 6],
          15,
          -1560198380
        ),
        s,
        n,
        e[l + 13],
        21,
        1309151649
      ),
      i = F(
        i,
        s = F(s, n = F(n, o, i, s, e[l + 4], 6, -145523070), o, i, e[l + 11], 10, -1120210379),
        n,
        o,
        e[l + 2],
        15,
        718787259
      ),
      s,
      n,
      e[l + 9],
      21,
      -343485551
    ), n = pe(n, r), o = pe(o, a), i = pe(i, u), s = pe(s, p);
  }
  return [n, o, i, s];
}
function At(e, t, n, o, i, s) {
  return pe(bi(pe(pe(t, e), pe(o, s)), i), n);
}
function N(e, t, n, o, i, s, l) {
  return At(t & n | ~t & o, e, t, i, s, l);
}
function B(e, t, n, o, i, s, l) {
  return At(t & o | n & ~o, e, t, i, s, l);
}
function _(e, t, n, o, i, s, l) {
  return At(t ^ n ^ o, e, t, i, s, l);
}
function F(e, t, n, o, i, s, l) {
  return At(n ^ (t | ~o), e, t, i, s, l);
}
function pe(e, t) {
  const n = (65535 & e) + (65535 & t);
  return (e >> 16) + (t >> 16) + (n >> 16) << 16 | 65535 & n;
}
function bi(e, t) {
  return e << t | e >>> 32 - t;
}
const So = Symbol("INVALID_KEY");
function Ii(e) {
  const t = e.slice(0, 32), n = e.slice(32), o = atob(n);
  if (t !== fi(n))
    return So;
  const [i, s] = o.split(";"), l = i.slice(2), r = new Date(Number(s.slice(2)));
  return { orderNumber: l, expiryDate: r };
}
const vi = {
  valid: !1,
  consoleMessage: "The VirtuosoMessageList license wrapper component is missing. Enclose the VirtuosoMessageList with VirtuosoMessageListLicense and add your key at the lisenceKey property.",
  watermarkMessage: "The VirtuosoMessageList license wrapper component is missing. Enclose the VirtuosoMessageList with VirtuosoMessageListLicense and add your key at the lisenceKey property."
}, Si = {
  valid: !1,
  consoleMessage: "Your VirtuosoMessageListLicense is missing a license key. Purchase one from https://virtuoso.dev/pricing/",
  watermarkMessage: "Your VirtuosoMessageListLicense is missing a license key. Purchase one from https://virtuoso.dev/pricing/"
}, ki = {
  valid: !1,
  consoleMessage: "Your VirtuosoMessageListLicense component is missing a license key - this component will not work if deployed in production. Purchase a key from https://virtuoso.dev/pricing/ before you deploy to production."
}, ko = {
  valid: !0
}, xi = {
  valid: !1,
  consoleMessage: "Your Virtuoso Message List license key is invalid. Ensure that you have copy-pasted the key from the purchase email correctly.",
  watermarkMessage: "Your Virtuoso Message List license key is invalid"
}, Ti = {
  valid: !1,
  consoleMessage: "Your annual license key to use Virtuoso Message List in non-production environments has expired. You can still use it in production. To keep using it in development, purchase a new key from https://virtuoso.dev/pricing/",
  watermarkMessage: "Your annual license key to use Virtuoso Message List in non-production environments has expired. You can still use it in production. To keep using it in development, purchase a new key from https://virtuoso.dev/pricing/"
}, Ei = {
  valid: !1,
  consoleMessage: "You have installed a version of `@virtuoso.dev/message-list` that is newer than the period of your license key. Either downgrade to a supported version, or purchase a new license from https://virtuoso.dev/pricing/",
  watermarkMessage: "You have installed a version of `@virtuoso.dev/message-list` that is newer than the period of your license key. Either downgrade to a supported version, or purchase a new license from https://virtuoso.dev/pricing/"
}, yi = ko, $i = /^(?:127\.0\.0\.1|localhost|0\.0\.0\.0|.+\.local)$/, wi = ["virtuoso.dev", "csb.app", "codesandbox.io"];
function Li({ licenseKey: e, now: t, hostname: n, packageTimestamp: o }) {
  const i = n.match($i), s = wi.some((r) => n.endsWith(r));
  if (!e)
    return s ? yi : i ? ki : Si;
  const l = Ii(e);
  if (l === So)
    return xi;
  if (l.expiryDate.getTime() < t.getTime()) {
    if (i)
      return Ti;
    if (l.expiryDate.getTime() < o)
      return Ei;
  }
  return ko;
}
const xo = h.createContext(vi), Ri = ({
  licenseKey: e,
  children: t
}) => {
  const n = Li({
    licenseKey: e,
    hostname: typeof window < "u" ? window.location.hostname : "localhost",
    now: /* @__PURE__ */ new Date(),
    packageTimestamp: 1755842099641
  });
  return /* @__PURE__ */ y(xo.Provider, { value: n, children: t });
};
Ri.displayName = Math.random().toString(36).slice(2, 8);
const Ot = h.createContext(void 0);
let kn = !1;
const Ci = h.forwardRef(
  ({
    initialData: e = [],
    computeItemKey: t = mo,
    context: n = null,
    initialLocation: o = null,
    shortSizeAlign: i = "top",
    onScroll: s,
    onRenderedDataChange: l,
    ItemContent: r = go,
    Header: a = null,
    StickyHeader: u = null,
    Footer: p = null,
    StickyFooter: d = null,
    EmptyPlaceholder: I = null,
    HeaderWrapper: b = lo,
    StickyHeaderWrapper: f = uo,
    FooterWrapper: m = ro,
    StickyFooterWrapper: L = co,
    useWindowScroll: q = !1,
    customScrollParent: U = null,
    ScrollElement: j = "div",
    increaseViewportBy: be = 0,
    data: ee,
    itemIdentity: ut = en,
    ...se
  }, W) => {
    const C = h.useMemo(() => {
      const R = new Lo();
      return R.register(Me), R.register(Mt), R.register($t), R.register(kt), R.register(bt), R.register(wt), R.register(oe), R.register(Ne), R.pubIn({
        [S]: e.slice(),
        [Io]: en,
        [He]: n,
        [bo]: t,
        [Fe]: o,
        [Qt]: r,
        [eo]: a,
        [no]: p,
        [to]: u,
        [oo]: d,
        [io]: I,
        [so]: j,
        [ho]: L,
        [po]: f,
        [fo]: m,
        [ao]: b,
        [tn]: i,
        [fn]: q,
        [Qe]: U,
        [Kn]: be
      }), R.singletonSub(Xt, s), R.singletonSub(Lt, l), R;
    }, []);
    h.useImperativeHandle(W, () => Qn(C), [C]), h.useEffect(() => {
      C.pubIn({
        [He]: n,
        [Qt]: r,
        [Qe]: U,
        [vo]: ee
      }), C.singletonSub(Xt, s), C.singletonSub(Lt, l);
    }, [n, r, U, s, l, C, ee]);
    const O = h.useContext(xo);
    return h.useEffect(() => {
      O.consoleMessage && (kn || (kn = !0, console.warn(O.consoleMessage)));
    }, [O]), h.useEffect(() => {
      const R = (te) => {
        var Ve;
        (Ve = te.message) != null && Ve.includes("ResizeObserver loop") && (te.preventDefault(), te.stopPropagation(), te.stopImmediatePropagation());
      };
      return window.addEventListener("error", R, { capture: !0 }), () => {
        window.removeEventListener("error", R);
      };
    }, []), typeof window < "u" && O.watermarkMessage ? /* @__PURE__ */ y(
      "div",
      {
        style: {
          color: "red",
          pointerEvents: "none"
        },
        children: O.watermarkMessage
      }
    ) : /* @__PURE__ */ y(Ro.Provider, { value: C, children: /* @__PURE__ */ y(Mi, { ...se }) });
  }
);
Ci.displayName = "VirtuosoMessageList";
const Mi = ({ style: e, ...t }) => {
  const n = we(), o = h.useContext(Ot), [
    i,
    s,
    l,
    r,
    a,
    u,
    p,
    d,
    I,
    b,
    f
  ] = Co(
    eo,
    to,
    ao,
    po,
    no,
    oo,
    fo,
    ho,
    Qt,
    io,
    Qe
  ), [m] = h.useState(() => {
    if (typeof window < "u" && typeof ResizeObserver > "u")
      throw new Error("ResizeObserver not found. Please ensure that you have a polyfill installed.");
    if (!(typeof ResizeObserver > "u"))
      return new ResizeObserver((x) => {
        var ft, Oe, Ye, gn;
        const ve = x.length, z = [];
        let $ = {};
        for (let Bt = 0; Bt < ve; Bt++) {
          const ce = x[Bt], A = ce.target;
          if (A === ee.current) {
            $ = {
              ...$,
              [it]: ce.contentRect.height,
              [D]: (ft = W.current) == null ? void 0 : ft.scrollHeight
            };
            continue;
          }
          if (A === se.current) {
            $ = {
              ...$,
              [ze]: ce.contentRect.height,
              [D]: (Oe = W.current) == null ? void 0 : Oe.scrollHeight
            };
            continue;
          }
          if (A === q.current) {
            $ = {
              ...$,
              [cn]: ce.contentRect.height,
              [D]: (Ye = W.current) == null ? void 0 : Ye.scrollHeight
            };
            continue;
          }
          if (A === j.current) {
            $ = {
              ...$,
              [st]: ce.contentRect.height,
              [D]: (gn = W.current) == null ? void 0 : gn.scrollHeight
            };
            continue;
          }
          if (A === W.current) {
            $ = {
              ...$,
              [w]: A.scrollTop,
              [D]: A.scrollHeight,
              [K]: ce.contentRect.height,
              [xt]: A.clientWidth
            };
            continue;
          }
          if (A === O.current) {
            W.current && ($ = {
              ...$,
              [D]: W.current.scrollHeight
            });
            continue;
          }
          if (A === C.current) {
            const ae = A.ownerDocument.defaultView;
            ae !== null && ($ = {
              ...$,
              [D]: ce.contentRect.height,
              [me]: ae.scrollY,
              [Pe]: ae.innerHeight,
              [et]: A.getBoundingClientRect().top + ae.scrollY,
              [xt]: A.clientWidth
            });
            continue;
          }
          if (A === f || A === R.current) {
            const ae = te.current, ht = R.current;
            ae && ht && ($ = {
              ...$,
              [D]: ht.getBoundingClientRect().height,
              [me]: ae.scrollTop,
              [Pe]: ae.clientHeight,
              [et]: ht.offsetTop,
              [xt]: ht.clientWidth
            });
            continue;
          }
          if (A.dataset.index === void 0)
            continue;
          const _t = Number.parseInt(A.dataset.index), wo = Number.parseFloat(A.dataset.knownSize ?? ""), Ft = ce.contentRect.height;
          if (Ft === wo)
            continue;
          const mn = z[z.length - 1];
          z.length === 0 || mn.size !== Ft || mn.endIndex !== _t - 1 ? z.push({ endIndex: _t, size: Ft, startIndex: _t }) : z[z.length - 1].endIndex++;
        }
        z.length > 0 && ($ = {
          ...$,
          [fe]: z
        }), n.pubIn($);
      });
  }), [L, q] = gt(m), [U, j] = gt(m), [be, ee] = gt(m), [ut, se] = gt(m), W = h.useRef(null), C = h.useRef(null), O = h.useRef(null), R = h.useRef(null), te = h.useRef(null);
  Tn(() => {
    te.current = f;
  }, [f]);
  const Ve = h.useCallback(
    (x) => {
      if (o) {
        const ve = Number.parseInt(x.dataset.index ?? "");
        n.pub(fe, [
          {
            startIndex: ve,
            endIndex: ve,
            size: o.itemHeight
          }
        ]);
      }
      m == null || m.observe(x);
    },
    [m, n, o]
  ), ct = h.useCallback(
    (x) => {
      m == null || m.unobserve(x);
    },
    [m]
  ), ne = h.useCallback(
    (x) => {
      x ? (O.current = x, m == null || m.observe(x, { box: "border-box" })) : O.current && (m == null || m.unobserve(O.current), O.current = null);
    },
    [m]
  ), { items: Ie } = M(Me), le = h.useCallback(() => {
    var ve;
    const x = [];
    for (const z of ((ve = O.current) == null ? void 0 : ve.children) ?? []) {
      if (z.dataset.index === void 0)
        continue;
      const $ = Number.parseInt(z.dataset.index), ft = Number.parseFloat(z.dataset.knownSize ?? ""), Oe = z.getBoundingClientRect().height;
      if (Oe === ft)
        continue;
      const Ye = x[x.length - 1];
      x.length === 0 || Ye.size !== Oe || Ye.endIndex !== $ - 1 ? x.push({ endIndex: $, size: Oe, startIndex: $ }) : x[x.length - 1].endIndex++;
    }
    n.pub(fe, x);
  }, [n]);
  h.useLayoutEffect(() => n.sub(Jn, le), [le, n]);
  const re = M(Y), Dt = M(Et), ue = M(Rt), Nt = M(Vt), To = M(Zn), Ae = M(He), Eo = M(bo), at = M(Ce), yo = M(We), $o = M(fn), pt = M(ke);
  return h.useLayoutEffect(() => {
    Ie.length === 0 && n.pub(Xn);
  }, [Ie, n]), h.useLayoutEffect(() => {
    var x;
    at > 0 && pt === null && b !== null && n.pub(D, (x = W.current) == null ? void 0 : x.scrollHeight);
  }, [n, at, pt, b]), /* @__PURE__ */ En(
    f ? Oi : $o ? Ai : Vi,
    {
      ...t,
      observer: m,
      scrollerRef: W,
      customScrollParentWrapperRef: R,
      listRef: O,
      style: e,
      windowScrollWrapperRef: C,
      children: [
        (at === 0 || pt) && b ? /* @__PURE__ */ y(b, { context: Ae }) : null,
        s && /* @__PURE__ */ y(r, { ref: ut, style: dt, children: /* @__PURE__ */ y(s, { context: Ae }) }),
        i && /* @__PURE__ */ y(l, { ref: be, style: dt, children: /* @__PURE__ */ y(i, { context: Ae }) }),
        at > 0 ? /* @__PURE__ */ y(
          "div",
          {
            ref: ne,
            "data-testid": "virtuoso-list",
            style: {
              boxSizing: "content-box",
              height: yo,
              paddingBottom: Nt,
              overflowAnchor: "none",
              marginTop: ue,
              transition: To,
              position: "relative",
              transform: `translateY(${re + Dt}px)`,
              visibility: pt ? "hidden" : "visible"
            },
            children: Ie.map((x) => /* @__PURE__ */ y(
              ci,
              {
                mount: Ve,
                unmount: ct,
                item: x,
                ItemContent: I
              },
              Eo({ index: x.index, data: x.data, context: Ae })
            ))
          }
        ) : null,
        a && /* @__PURE__ */ y(p, { ref: L, style: dt, children: /* @__PURE__ */ y(a, { context: Ae }) }),
        u && /* @__PURE__ */ y(d, { ref: U, style: dt, children: /* @__PURE__ */ y(u, { context: Ae }) })
      ]
    }
  );
}, Vi = ({
  customScrollParentWrapperRef: e,
  windowScrollWrapperRef: t,
  observer: n,
  children: o,
  listRef: i,
  scrollerRef: s,
  style: l,
  ...r
}) => {
  const a = we(), u = h.useContext(Ot), p = M(so), { onScroll: d, onWheel: I } = hn({
    listRef: i,
    scrollTopCell$: w,
    scrollToSignal$: $e,
    scrollableRef: s
  }), b = h.useCallback(
    (L) => {
      L ? (a.pub(ye, L), s.current = L, L.addEventListener("scroll", d), L.addEventListener("wheel", I), u && a.pubIn({
        [K]: u.viewportHeight,
        [D]: u.viewportHeight,
        [w]: 0
      }), n == null || n.observe(L, { box: "border-box" })) : s.current && (s.current.removeEventListener("scroll", d), s.current.removeEventListener("wheel", I), a.pub(ye, null), n == null || n.unobserve(s.current), s.current = null);
    },
    [n, a, d, I, u, s]
  );
  dn(() => {
    var L;
    return (L = s.current) == null ? void 0 : L.scrollHeight;
  });
  const f = M(G), m = M(He);
  return /* @__PURE__ */ y(
    p,
    {
      ...r,
      ref: b,
      "data-testid": "virtuoso-scroller",
      style: {
        overflowY: f ? "hidden" : "scroll",
        boxSizing: "border-box",
        ...l
      },
      ...p === "div" ? {} : { context: m },
      children: o
    }
  );
}, Ai = ({
  observer: e,
  children: t,
  windowScrollWrapperRef: n,
  customScrollParentWrapperRef: o,
  listRef: i,
  ...s
}) => {
  const l = h.useRef(null), r = we(), a = h.useContext(Ot), { onScroll: u, onWheel: p } = hn({
    listRef: i,
    scrollTopCell$: me,
    scrollToSignal$: an,
    scrollableRef: l
  }), d = h.useCallback(() => {
    var f;
    const b = l.current;
    b !== null && r.pub(Pe, (f = b.ownerDocument.defaultView) == null ? void 0 : f.innerHeight);
  }, [r]), I = h.useCallback(
    (b) => {
      if (b) {
        r.pub(ye, b), n.current = b;
        const f = b.ownerDocument.defaultView;
        f && (f.addEventListener("scroll", u), f.addEventListener("wheel", p), f.addEventListener("resize", d), a && r.pubIn({
          [Pe]: f == null ? void 0 : f.innerHeight,
          [D]: b.getBoundingClientRect().height,
          [me]: 0
        }), l.current = b.ownerDocument.documentElement), e == null || e.observe(b, { box: "border-box" });
      } else {
        if (n.current) {
          const f = n.current.ownerDocument.defaultView;
          f && (f.removeEventListener("scroll", u), f.removeEventListener("wheel", p), f.removeEventListener("resize", d)), r.pub(ye, null), e == null || e.unobserve(n.current), n.current = null;
        }
        l.current = null;
      }
    },
    [e, r, u, p, d, a, n]
  );
  return dn(() => {
    var b;
    return (b = n.current) == null ? void 0 : b.getBoundingClientRect().height;
  }), /* @__PURE__ */ y("div", { ref: I, ...s, children: t });
}, Oi = ({
  scrollerRef: e,
  windowScrollWrapperRef: t,
  children: n,
  listRef: o,
  customScrollParentWrapperRef: i,
  observer: s,
  ...l
}) => {
  const r = we(), a = h.useContext(Ot), u = M(Qe), p = h.useRef(u);
  Tn(() => {
    p.current = u;
  }, [u]);
  const { onWheel: d, onScroll: I } = hn({
    listRef: o,
    scrollTopCell$: me,
    scrollToSignal$: an,
    scrollableRef: p
  }), b = h.useCallback(
    (f) => {
      if (f) {
        const m = p.current;
        r.pub(ye, m), i.current = f, m && (m.addEventListener("scroll", I), m.addEventListener("wheel", d), a && r.pubIn({
          [Pe]: m.clientHeight,
          [D]: f.getBoundingClientRect().height,
          [me]: 0
        }), s == null || s.observe(m, { box: "border-box" })), s == null || s.observe(f, { box: "border-box" });
      } else {
        const m = p.current;
        m && (m.removeEventListener("scroll", I), m.removeEventListener("wheel", d), r.pub(ye, null), s == null || s.unobserve(m)), i.current && (s == null || s.unobserve(i.current)), i.current = null;
      }
    },
    [s, r, I, d, a, i]
  );
  return dn(() => {
    var f;
    return (f = i.current) == null ? void 0 : f.getBoundingClientRect().height;
  }), /* @__PURE__ */ y("div", { ref: b, ...l, children: n });
};
function hn({ scrollToSignal$: e, scrollableRef: t, listRef: n, scrollTopCell$: o }) {
  const i = we(), s = h.useRef(null), l = pi(t, n, s), r = h.useCallback(
    (p) => {
      t.current && (t.current.scrollTop += p);
    },
    [t]
  ), a = h.useCallback(() => {
    const p = t.current;
    if (p !== null && (i.pub(o, p.scrollTop), s.current !== null)) {
      const d = p.scrollHeight - p.clientHeight;
      Bn(p.scrollTop, Math.min(d, s.current)) && (s.current = null, i.pub(un, !1), i.pub(ot, p.scrollTop));
    }
  }, [i, t, o]), u = h.useCallback(
    (p) => {
      i.pub(pn, p.deltaY > 0 ? "down" : "up");
    },
    [i]
  );
  return h.useLayoutEffect(() => i.sub(e, l), [l, i, e]), h.useLayoutEffect(() => i.sub(X, r), [r, i]), {
    onScroll: a,
    onWheel: u
  };
}
function dn(e) {
  const t = Mo(D);
  Vo(() => {
    if (!_n())
      return;
    const n = setInterval(() => {
      t(e() ?? 0);
    }, 1e3);
    return () => {
      clearInterval(n);
    };
  }, [t, e]);
}
export {
  Wi as ScrollModifierOption,
  Ci as VirtuosoMessageList,
  Ri as VirtuosoMessageListLicense,
  Ot as VirtuosoMessageListTestingContext,
  Hi as useCurrentlyRenderedData,
  Fi as useVirtuosoLocation,
  Pi as useVirtuosoMethods
};
